x/mongo/driver/topology/topology_options.go

```diff
@@ -7,451 +7,370 @@
-		}
+// convertToDriverAPIOptions converts a options.ServerAPIOptions instance to a driver.ServerAPIOptions.
+func convertToDriverAPIOptions(s *options.ServerAPIOptions) *driver.ServerAPIOptions {
+	driverOpts := driver.NewServerAPIOptions(string(s.ServerAPIVersion))
+	if s.Strict != nil {
+		driverOpts.SetStrict(*s.Strict)
 	}
-
-	return cfg, nil
+	if s.DeprecationErrors != nil {
+		driverOpts.SetDeprecationErrors(*s.DeprecationErrors)
+	}
+	return driverOpts
 }
 
-// WithConnString configures the topology using the connection string.
-func WithConnString(fn func(connstring.ConnString) connstring.ConnString) Option {
-	return func(c *config) error {
-		cs := fn(c.cs)
-		c.cs = cs
-
-		if cs.ServerSelectionTimeoutSet {
-			c.serverSelectionTimeout = cs.ServerSelectionTimeout
-		}
-
-		var connOpts []ConnectionOption
-
-		if cs.AppName != "" {
-			c.serverOpts = append(c.serverOpts, WithServerAppName(func(string) string { return cs.AppName }))
-		}
-
-		if cs.Connect == connstring.SingleConnect || (cs.DirectConnectionSet && cs.DirectConnection) {
-			c.mode = SingleMode
-		}
+type client interface {
+	GetClusterClock() *session.ClusterClock
+	SetServerAPI(*driver.ServerAPIOptions)
+}
 
-		c.seedList = cs.Hosts
+// newConfig will translate data from client options into a topology config for building non-default deployments.
+func newConfig(co *options.ClientOptions, client client) (*Config, error) {
+	var defaultMaxPoolSize uint64 = 100
+	var serverAPI *driver.ServerAPIOptions = nil
 
-		if cs.ConnectTimeout > 0 {
-			c.serverOpts = append(c.serverOpts, WithHeartbeatTimeout(func(time.Duration) time.Duration { return cs.ConnectTimeout }))
-			connOpts = append(connOpts, WithConnectTimeout(func(time.Duration) time.Duration { return cs.ConnectTimeout }))
-		}
+	var clusterClock *session.ClusterClock = nil
+	if client != nil {
+		clusterClock = client.GetClusterClock()
+	}
 
-		if cs.SocketTimeoutSet {
-			connOpts = append(
-				connOpts,
-				WithReadTimeout(func(time.Duration) time.Duration { return cs.SocketTimeout }),
-				WithWriteTimeout(func(time.Duration) time.Duration { return cs.SocketTimeout }),
-			)
-		}
+	var defaultOptions int
+	// Set default options
+	if co.MaxPoolSize == nil {
+		defaultOptions++
+		co.SetMaxPoolSize(defaultMaxPoolSize)
+	}
+	if err := co.Validate(); err != nil {
+		return nil, err
+	}
 
-		if cs.HeartbeatInterval > 0 {
-			c.serverOpts = append(c.serverOpts, WithHeartbeatInterval(func(time.Duration) time.Duration { return cs.HeartbeatInterval }))
-		}
+	var connOpts []ConnectionOption
+	var serverOpts []ServerOption
 
-		if cs.MaxConnIdleTime > 0 {
-			connOpts = append(connOpts, WithIdleTimeout(func(time.Duration) time.Duration { return cs.MaxConnIdleTime }))
-		}
+	cfgp := new(Config)
 
-		if cs.MaxPoolSizeSet {
-			c.serverOpts = append(c.serverOpts, WithMaxConnections(func(uint64) uint64 { return cs.MaxPoolSize }))
-		}
+	// TODO(GODRIVER-814): Add tests for topology, server, and connection related options.
```

Comment:
[GODRIVER-814](https://jira.mongodb.org/browse/GODRIVER-814) is closed and this comment seems outdated, so can be removed.